pkg/host/cpu.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and IronCore contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package host
+
+import (
+	"encoding/xml"
+	"fmt"
+
+	"github.com/digitalocean/go-libvirt"
+	"libvirt.org/go/libvirtxml"
+)
+
+// CpuTopology return a list of related to NUMA node CPU cores
+func CpuTopology(lv *libvirt.Libvirt) (map[int][]int, error) {
+	capsData, err := lv.Capabilities()
+	if err != nil {
+		return nil, fmt.Errorf("error getting capabilities: %w", err)
+	}
+
+	var caps libvirtxml.Caps
+	if err := xml.Unmarshal(capsData, &caps); err != nil {
+		return nil, fmt.Errorf("error unmarshalling guest capabilities: %w", err)
+	}
+	if caps.Host.NUMA == nil || caps.Host.NUMA.Cells == nil || len(caps.Host.NUMA.Cells.Cells) == 0 {
+		return nil, fmt.Errorf("no NUMA node detected on the host")
+	}
+
+	node := make(map[int][]int, caps.Host.NUMA.Cells.Num)
+
+	for _, cell := range caps.Host.NUMA.Cells.Cells {
+		cpus := make([]int, cell.CPUS.Num)
+		for i, c := range cell.CPUS.CPUs {
+			cpus[i] = c.ID
+		}
+		node[cell.ID] = cpus
+	}
+	return node, nil
+}
+
+// getVcpuPinInfo return the CPU affinity setting of all virtual CPUs of domain
+func getVcpuPinInfo(lv *libvirt.Libvirt, domain libvirt.Domain) ([][]bool, error) {
+	_, _, _, _, rNodes, rSockets, rCores, rThreads, err := lv.NodeGetInfo()
+	if err != nil {
+		return nil, err
+	}
+	npcpus := rNodes * rSockets * rCores * rThreads
+	maplen := (npcpus + 7) / 8
```

Comment:
why 7 / 8? what are these numbers?

Reply:
It is a common idiom to calculate the byte length needed to represent a bitmap or array where each bit corresponds to a specific CPU, ensuring that there is enough space for rounding

Reply:
8 is count of bits for byte
what is 7?

Reply:
it's to ensure that there is enough space for rounding.